File: Python/1043-is_it_a_binary_search_tree.py
import sys
sys.setrecursionlimit(1024)


def postorder(pre, post, is_mirror):
    if not pre:
        return

    root = pre[0]
    l_pre, r_pre = [], []
    for node in pre[1:]:
        if is_mirror:
            if node >= root:
                if r_pre:
                    break
                else:
                    l_pre.append(node)
            else:
                r_pre.append(node)
        else:
            if node < root:
                if r_pre:
                    break
                else:
                    l_pre.append(node)
            else:
                r_pre.append(node)

    postorder(l_pre, post, is_mirror)
    postorder(r_pre, post, is_mirror)
    post.append(root)


if __name__ == '__main__':
    _ = input()
    pre = [int(i) for i in input().split()]
    post = []
    postorder(pre, post, is_mirror=False)
    if len(post) != len(pre):
        post = []
        postorder(pre, post, is_mirror=True)
    if len(post) == len(pre):
        print("YES")
        print(*post)
    else:
        print("NO")

File: C++/1043-is_it_a_binary_search_tree.cc
#include <iostream>
#include <vector>


std::vector<int> pre;
bool is_mirror = false;


void
postorder(std::size_t bgn, std::size_t end, std::vector<int> &post)
{
    if (bgn == end) return;

    int root = pre[bgn];
    std::size_t l_end = bgn + 1, r_bgn = end - 1;

    if (!is_mirror) {
        while (l_end < end && pre[l_end] < root) ++l_end;
        while (r_bgn > bgn && pre[r_bgn] >= root) --r_bgn;
    } else {
        while (l_end < end && pre[l_end] >= root) ++l_end;
        while (r_bgn > bgn && pre[r_bgn] < root) --r_bgn;
    }

    postorder(bgn + 1, l_end, post);
    postorder(r_bgn + 1, end, post);
    post.push_back(root);
}


int main() {

    std::ios::sync_with_stdio(false);

    std::size_t N;
    std::cin >> N;

    for (std::size_t i=0; i<N; ++i) {
        int node;
        std::cin >> node;
        pre.push_back(node);
    }

    std::vector<int> post;
    postorder(0, N, post);
    if (post.size() != N) {
        post.clear();
        is_mirror = true;
        postorder(0, N, post);
    }
    if (post.size() == N) {
        std::cout << "YES\n";
        for (auto iter=post.begin(); iter!=post.end()-1; ++iter) {
            std::cout << *iter << ' ';
        }
        std::cout << post.back() << '\n';
    } else {
        std::cout << "NO\n";
    }

    return 0;
}
